Sycl Backend Update

homerdin opened the pull request:
# Summary

This PR includes initial backend support for SYCL kernels, foralls and reductions.  These features were discussed in #1010.  There is also an initial pass at the documentation.

# Design review

There are several open issues for the SYCL backend.

- non-trivially copyable kernels.  This PR provides the `SyclKernelNonTrivial` and `sycl_exec_nontrivial` policies to support non-trivially copyable kernels.  This will work if the kernel can be memcopied to the device.  This is a workaround until  the SYCL 2020 feature `is_device_copyable` type trait is supported.  This features is described in the spec at `4.12.3` and `3.13.1`.  

- Work group size limitations - Available work groups sizes for kernel execution depend on the device.  We have the option to query the device to guarantee valid work group sizes, but it might need to change what is specified in the policies.  Currently if the policies specify an invalid work group size, we will see a runtime error about the invalid work group size.

- Execution resource workaround - SYCL cannot query the runtime to determine which device is currently being used, the device is tied to the `sycl::queue` object.  This matches the new resource interface well, but to support the other interface we've added a global pointer in RAJA to know what resource should be used.

- Reductions - The reduction implementation is not very optimized currently.  Ideally we would leverage the SYCL `reduction` directly.  There are extensions which allow access to the location in the iterate space without access to the `nd_item` which can be leveraged to improve the current reductions.  These extensions are not currently portable.

rhornung67 commented:
Folks, please start looking over this PR so we can include it in the release. @davidbeckingsale @trws @ajkunen @MrBurmark @artv3 @rchen20 @mdavis36

artv3 commented:
Could we add a couple of examples? Like add vectors and dot product?

artv3 commented on include/RAJA/policy/sycl.hpp:
```diff
@@ -0,0 +1,42 @@
+//#include "RAJA/policy/cuda/sort.hpp"
+#include "RAJA/policy/sycl/kernel.hpp"
+//#include "RAJA/policy/cuda/synchronize.hpp"
+//#include "RAJA/policy/cuda/WorkGroup.hpp"
```
If the commented code is not needed, it would be nice to remove.

artv3 replied:
Actually, I see what you are doing, perhaps change cuda->sycl.

artv3 commented on include/RAJA/policy/sycl/reduce.hpp:
```diff
@@ -0,0 +1,511 @@
+  {
+    if (!host) {
+      printf("Unable to allocate space on host\n");
+      exit(1);
```
Consider using RAJA_ABORT_OR_THROW

artv3 commented:
What about adding some functional test as well? I believe it would just be a matter of adding the SYCL policies here: 
https://github.com/LLNL/RAJA/blob/develop/test/include/RAJA_test-forall-execpol.hpp
and adding them to the various functional test.

MrBurmark commented on include/RAJA/policy/sycl/forall.hpp:
```diff
@@ -0,0 +1,279 @@
+    lbody = (LOOP_BODY*) cl::sycl::malloc_device(sizeof(LOOP_BODY), *q);
+    q->memcpy(lbody, &loop_body, sizeof(LOOP_BODY)).wait();
+
+    beg = (Iterator*) cl::sycl::malloc_device(sizeof(Iterator), *q);
+    q->memcpy(beg, &begin, sizeof(Iterator)).wait();
```
Why copy the iterator as well?

MrBurmark commented on include/RAJA/policy/sycl/kernel/For.hpp:
```diff
@@ -0,0 +1,478 @@
+    auto i_stride = item.get_local_range(Dim);
+    auto i = i0;
+
+    for(; i < len;i += i_stride){
+
+      // Assign the x thread to the argument
+      data.template assign_offset<ArgumentId>(i);
+
+      // execute enclosed statements
+      enclosed_stmts_t::exec(data, item, thread_active);
+    }
+    // do we need one more masked iteration?
+    if(i - i0 < len)
+    {
+      // execute enclosed statements one more time, but masking them off
+      // this is because there's at least one thread that isn't masked off
+      // that is still executing the above loop
+      enclosed_stmts_t::exec(data, item, false);
+    }
```
I'm pretty sure this was changed in the cuda impl to use a single loop so the threads execute the same line of code at the same time as a way of avoiding divergence.

MrBurmark commented on include/RAJA/policy/sycl/kernel/SyclKernel.hpp:
```diff
@@ -0,0 +1,296 @@
+ */
+template <typename LaunchConfig, typename... EnclosedStmts>
+struct SyclKernelExtNonTrivial
+    : public internal::Statement<sycl_exec<0>, EnclosedStmts...> {
```
Should this be sycl_launch instead of sycl_exec?

MrBurmark commented on include/RAJA/policy/sycl/kernel/SyclKernel.hpp:
```diff
@@ -0,0 +1,296 @@
+ */
+template <typename LaunchConfig, typename... EnclosedStmts>
+struct SyclKernelExt
+    : public internal::Statement<sycl_exec<0>, EnclosedStmts...> {
```
same as above

MrBurmark commented on include/RAJA/policy/sycl/kernel/SyclKernel.hpp:
```diff
@@ -0,0 +1,296 @@
+ * This is annotated to guarantee that device code generated
+ * can be launched by a kernel with BlockSize number of threads.
+ *
+ * This launcher is used by the SyclKerelFixed policies.
```
typo

MrBurmark commented on include/RAJA/policy/sycl/kernel/internal.hpp:
```diff
@@ -0,0 +1,217 @@
+
+  RAJA_INLINE
+  RAJA_HOST_DEVICE
+  LaunchDims(LaunchDims const &c) :
+  group(c.group),//   min_group(c.min_group),
+  local(c.local),//, min_local(c.min_local)
+  global(c.global)
```
Confusing indentation.

MrBurmark commented on include/RAJA/policy/sycl/kernel/internal.hpp:
```diff
@@ -0,0 +1,217 @@
+
+    // Note: Work group allowable sizes depend on the device
+    //       Could query the device to set them
+    //       For now, error on bad work group sizei
```
typo

MrBurmark commented on include/RAJA/policy/sycl/policy.hpp:
```diff
@@ -0,0 +1,234 @@
+}  // namespace sycl
+}  // namespace policy
+
+using policy::sycl::sycl_exec;
+using policy::sycl::sycl_exec_nontrivial;
```
Is it possible to use the compiler to make the trivial/non-trivial differentiation automatically?

rhornung67 replied:
That would be most excellent, if possible.

homerdin replied:
Yes, I had previously tried to wrap to two different types of launches in an `if constexpr` using `std::is_trivially_copyable`.  While it works on a small reproducer, when used in RAJA is it triggering a bug (tracking it).  I haven't tried to implement using `std::enable_if`, I will try it.

rhornung67 replied:
Sounds good. If it works, it will be easier on users.  Please keep us posted.

MrBurmark commented on src/MemUtils_SYCL.cpp:
```diff
@@ -0,0 +1,77 @@
+syclInfo g_status;
+
+//! State of the host code in this thread
+syclInfo tl_status;
+#if defined(RAJA_ENABLE_OPENMP) && defined(_OPENMP)
+#pragma omp threadprivate(tl_status)
+#endif
```
Will this link with the symbol in the header if openmp is enabled?

trws replied:
Not correctly, is there a reason this needs to be `extern` rather than having functions that can retrieve or set it?

MrBurmark commented on include/RAJA/policy/sycl/forall.hpp:
```diff
@@ -0,0 +1,279 @@
+      h.parallel_for( cl::sycl::nd_range<1>{gridSize, blockSize},
+                      [=]  (cl::sycl::nd_item<1> it) {
+
+        size_t ii = it.get_global_id(0);
+        if (ii < len) {
```
Should ii be IndexType?

rhornung67 commented:
@homerdin how is this coming along? Anything we can help you with?

homerdin commented:
@rhornung67 Yeah thanks, I'm finishing up pushing several of the changes we discussed and addressing some of the comments.  Any additional comments or testing would be great.

Couple of points:
1. I've pushed some scripts to build using the DPCPP compiler with the cuda backend.  Hopefully this will help to enable some testing options. 
2. I've made the change to use the template resource in the forall launches.  It currently only uses the argument resource if the global resource (`setQueue`) is not set.  Another option is to check if the argument resource is the default resource and not use it (would only work if the application is using the default camp resource also for memory management).  Another alternative would be to have `setQueue` set the default resource, but this would make bigger changes in camp.

rhornung67 commented:
> @rhornung67 Yeah thanks, I'm finishing up pushing several of the changes we discussed and addressing some of the comments. Any additional comments or testing would be great.
> 
> Couple of points:
> 
>     1. I've pushed some scripts to build using the DPCPP compiler with the cuda backend.  Hopefully this will help to enable some testing options.
> 
>     2. I've made the change to use the template resource in the forall launches.  It currently only uses the argument resource if the global resource (`setQueue`) is not set.  Another option is to check if the argument resource is the default resource and not use it (would only work if the application is using the default camp resource also for memory management).  Another alternative would be to have `setQueue` set the default resource, but this would make bigger changes in camp.

@trws should weigh in on this.

trws commented:
> 2\. I've made the change to use the template resource in the forall launches.  It currently only uses the argument resource if the global resource (`setQueue`) is not set.  Another option is to check if the argument resource is the default resource and not use it (would only work if the application is using the default camp resource also for memory management).  Another alternative would be to have `setQueue` set the default resource, but this would make bigger changes in camp.

I'm not sure I'm understanding the consideration here.  We intentionally don't do the thread-local or global resource thing currently, why would it be preferable to do so here?  I have a feeling there's some dependency or setup issue that needs to be worked around here, and I'm not sure I know what the constraints are, why is setQueue important for the SYCL backend?

homerdin commented:
@trws It is mostly to ensure we don't end up with mis-matched contexts.  My understanding is that if the RAJA forall does not pass the resource argument, the default resource is passed.  Unless the application is using the context from the default resource for its memory management, the memory won't be usable.

rhornung67 commented on docs/sphinx/user_guide/feature/policies.rst:
```diff
@@ -398,6 +398,76 @@ Finally
+                                                        in a GPU kernel launched
+                                                        with given work group
+                                                        size.
+ sycl_global_1<WORK_GROUP_SIZE>           kernel (For)  Map loop iterates
```
@homerdin it looks like the sycl exec policies have been changed from sycl_*_{1,2,3} to sycl_*_{0,1,2}. Please update the polcy documentation here.  Thank you.

trws commented:
> @trws It is mostly to ensure we don't end up with mis-matched contexts. My understanding is that if the RAJA forall does not pass the resource argument, the default resource is passed. Unless the application is using the context from the default resource for its memory management, the memory won't be usable.

Got it, that's clearly something we need to deal with.  I had forgotten that there's no way to get a global default context provided by sycl.  Since we have demand for it on the other backends too, I think the first thing to do is have a way to create a camp resource from a sycl queue.   The second could be handled a couple of different ways, but I'm thinking the best way to deal with this is:

1. Have the sycl resource create all queues it uses from a single context
2. Store the array of queues in an array or map or some such indexed by context
3. Allow the user to override the default context by passing one to the constructor for a resource
4. Allow changing the default for a given thread with a setContext or similar much like the setQueue you have in here now

Result would be that if *everything* is using RAJA, the user doesn't need to care about alternate contexts.  If most things are using raja but there's some external sycl it's easy to get a set of queues that play nice with a context, and if it's a bit of raja that gets a queue a one-off resource can be created using a sycl queue.

I don't love having 4 in that list, but I think we almost have to for this since the defaults are likely to be incompatible in a wide variety of cases.  At least if it's by context rather than queue though, users don't get implicit sync with a particular external queue that way, which should help with the issues I was seeing with an older interface that allowed scoped ordering like that.

What do you think @homerdin?

rchen20 commented:
Hi @homerdin, I'm trying to build your fork on one of our machines, and ran in to this error:

/g/g20/chen59/work/RAJAsycl/raja_git_homerdinsycl/src/MemUtils_SYCL.cpp:63:48: error: no member named 'get_queue' in 'camp::resources::Sycl'
    return app_q->get<camp::resources::Sycl>().get_queue();
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^

I see get_stream and get_event were added to the CAMP SYCL backend. Is this maybe a typo, or were there some missing commits to CAMP?

homerdin commented:
@rchen20 Thanks for testing it out.  What repo are you pulling camp from?  The changes for get_queue are upstream (https://github.com/LLNL/camp.git) and also in my repo (https://github.com/homerdin/camp.git) on branch `sycl_be`.  I'm realizing that the default repo on my fork ing GitHub is pointing to an old branch in camp, I'll work on cleaning it up.

rchen20 commented:
Thanks @homerdin, checking out your camp/sycl_be branch fixed that error!

rhornung67 commented:
@homerdin please resolve the conflicts in your branch and make sure the latest RAJA develop is merged in so I can pull it into the RAJA repo for complete CI testing before merging.  Thank you.

homerdin commented:
@rhornung67 I've merged in develop and resolved the conflicts.  Please let me know if there are any issues.

rhornung67 commented:
Thanks @homerdin

rhornung67 commented:
@homerdin I merged the SYCL PR branch. Should I close this PR?

homerdin commented:
@rhornung67 That's great.  I'll close this and the older one.  Thanks
